agent/xds/server.go

```diff
@@ -0,0 +1,418 @@
+// to be written in the agent package to allow resolving without tightly
+// coupling this to the agent.
+type ACLResolverFunc func(id string) (acl.ACL, error)
+
+// ConnectAuthz is the interface the agent needs to expose to be able to re-use
+// the authorization logic between both APIs.
+type ConnectAuthz interface {
+	// ConnectAuthorize is implemented by Agent.ConnectAuthorize
+	ConnectAuthorize(token string, req *structs.ConnectAuthorizeRequest) (authz bool, reason string, m *cache.ResultMeta, err error)
+}
+
+// ConfigManager is the interface the config manager implements to make testing
+// the server without a full mocker manager, state and cache. In normal use this
+// is satisfied directly by a pointer to the agent's proxycfg.Manager.
+type ConfigManager interface {
+	Watch(proxyID string) (<-chan *proxycfg.ConfigSnapshot, func())
+}
+
+// Server represents a gRPC server that can handle both XDS and ext_authz
+// requests from Envoy.
+//
+// A full description of the XDS protocol can be found at
+// https://github.com/envoyproxy/data-plane-api/blob/master/XDS_PROTOCOL.md
+type Server struct {
+	logger       *log.Logger
+	cfgMgr       ConfigManager
+	authz        ConnectAuthz
+	resolveToken ACLResolverFunc
+}
+
+// NewServer creates a usable server instance.
+func NewServer(logger *log.Logger, cfgMgr ConfigManager,
+	aclFn ACLResolverFunc, authz ConnectAuthz) *Server {
+	return &Server{
+		logger:       logger,
+		cfgMgr:       cfgMgr,
+		authz:        authz,
+		resolveToken: aclFn,
+	}
+}
+
+// StreamAggregatedResources implements
+// discovery.AggregatedDiscoveryServiceServer. This is the ADS endpoint which is
+// the only xDS API we directly support for now.
+func (s *Server) StreamAggregatedResources(stream ADSStream) error {
+	// a channel for receiving incoming requests
+	reqCh := make(chan *v2.DiscoveryRequest)
+	reqStop := int32(0)
+	go func() {
+		for {
+			req, err := stream.Recv()
+			if atomic.LoadInt32(&reqStop) != 0 {
+				return
+			}
+			if err != nil {
+				close(reqCh)
+				return
+			}
+			reqCh <- req
+		}
+	}()
+
+	err := s.process(stream, reqCh)
+	if err != nil {
+		s.logger.Printf("[DEBUG] Error handling ADS stream: %s", err)
+	}
+
+	// prevents writing to a closed channel if send failed on blocked recv
+	atomic.StoreInt32(&reqStop, 1)
+
+	return err
+}
+
+const (
+	stateInit int = iota
+	statePendingAuth
+	stateRunning
+)
+
+func (s *Server) process(stream ADSStream, reqCh <-chan *v2.DiscoveryRequest) error {
+	// xDS requires a unique nonce to correlate response/request pairs
+	var nonce uint64
+
+	// xDS works with versions of configs. Internally we don't have a consistent
+	// version. We could just hash the config since versions don't have to be
+	// ordered as far as I can tell, but it's cheaper just to increment a counter
+	// every time we observe a new config since the upstream proxycfg package only
+	// delivers updates when there are actual changes.
+	var configVersion uint64
```

Comment:
How does this whole version dance work if Consul is restarted while Envoy remains running?

Reply:
Great question.

I think this is safe right now due to the way `xDSType.lastVersion` is handled lower down this file. I'll walk through an example to convince myself. (This is more by coincidence than good design.)

 - Assume Envoy and agent are up and Envoy has loaded version 3 already.
 - Agent restarts
 - Envoy reconnects with same discovery request showing it's current "version = 3"
 - Agent sees the current config snapshot after Watch was called which it assigns `version 1` due to having restarted
 - It delivers this
 - SendIfNew compares with `t.lastVersion` which is 0 since that was also just re-initialised and this is only the last version we _sent_ not updated by the incoming request
   - Since 1 > 0, the version is sent
 - Envoy applies the config and is now on `version 1` and resumes normally. 
 - We set t.lastVersion to be 1 and continue normally.

I think there is one corner cases where it might be problematic though:

If envoy is on version 1, the agent is stopped, the proxy config altered and then started again.
    - In this case the new config will be sent to Envoy but since it is also "version 1" Envoy _might_ ignore it as the same as the config it has applied (need to verify).

I think it's OK for now on the basis that I took this exact pattern from the xDS SDK and reference implementation. I assume based on that it's either not a real problem (because Envoy will re-apply config even if version is same as current) or is a huge edge case that's not a big deal yet. We can revisit later I guess.

Reply:
I tested it seems like Envoy still applies a config even if it's version is the same as the current one. In my case I started agent, let it deliver initial certs (at version 1) then restarted agent. When envoy reconnects it gets and loads a new set of certs even though it's still v1. So we are good for now.

Reply:
If you wanted more belt-and-suspenders you could either persist the last version to disk on the agent and pick up where you left off, or see if you could have the agent generate a random token on startup and see if there's a way to give that to the envoy process to strongly indicate "I'm controlled by a new master, so I should assume it's all different now". Sorta like a raft term.

Reply:
Or perhaps have the agent interrogate envoy to see what the loaded version is, and then internally sync the agent so that the next version it hands out is `envoyLoadedVersion+1`